src/include/api.h

```diff
@@ -63,10 +93,26 @@
     __oldname = (s)->name;                                                                   \
     ++(s)->api_call_counter;                                                                 \
     (s)->dhandle = (dh);                                                                     \
-    (s)->name = (s)->lastop = #struct_name "." #func_name
-#define API_SESSION_POP(s)  \
-    (s)->dhandle = __olddh; \
-    (s)->name = __oldname;  \
+    (s)->name = (s)->lastop = #struct_name "." #func_name;                                   \
+    if ((s)->id != 0 && (s)->api_call_counter == 1) {                                        \
```

Comment:
Reviewing this deserves some care: The caller says "No code before this line", I think this violates that. I think it's OK since the code never generates an error return, so the error setup should always complete, but it's worth taking some time to think through

Reply:
I agree this is infallible. Whether being infallible is enough to allow an exception to the "no code before this line" comment, I don't know.

Reply:
Thanks - that's enough